File: Release/Source/BBC.Dna/PageUI.cs


using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using BBC.Dna.Component;
using BBC.Dna;

namespace BBC.Dna
{
    /// <summary>
    /// The page UI class. Used to put the users page preferences into the current page
    /// </summary>
    public class PageUI : DnaComponent
    {
        ///TODO: replace with BBC.Dna.Objects.PageUi version
        /// <summary>
        /// Default constructor
        /// </summary>
        /// <param name="userID">The id of the user you want to get the page UI for</param>
        public PageUI(int userID)
        {
            // Create the basic layout for the XML
            XmlNode pageUI = AddElementTag(RootElement, "PAGEUI");

            // Add the Site home tag
            XmlNode current = AddElementTag(pageUI, "SITEHOME");
            AddAttribute(current, "VISIBLE", "1");
            AddAttribute(current, "LINKHINT", "/");

            // Add the search tag
            current = AddElementTag(pageUI, "SEARCH");
            AddAttribute(current, "VISIBLE", "1");
            AddAttribute(current, "LINKHINT", "/search");

            // Add the dont panic tag
            current = AddElementTag(pageUI, "DONTPANIC");
            AddAttribute(current, "VISIBLE", "1");
            AddAttribute(current, "LINKHINT", "/dontpanic");

            if (userID > 0)
            {
                // Add the My Home tag
                current = AddElementTag(pageUI, "MYHOME");
                AddAttribute(current, "VISIBLE", "1");
                AddAttribute(current, "LINKHINT", "/U" + userID.ToString());

                // Add the Register tag
                current = AddElementTag(pageUI, "REGISTER");
                AddAttribute(current, "VISIBLE", "1");

                // Add the My Details tag
                current = AddElementTag(pageUI, "MYDETAILS");
                AddAttribute(current, "VISIBLE", "1");
                AddAttribute(current, "LINKHINT", "/UserDetails");

                // Add the Logout tag
                current = AddElementTag(pageUI, "LOGOUT");
                AddAttribute(current, "VISIBLE", "1");
                AddAttribute(current, "LINKHINT", "/Logout");
            }
            else
            {
                // Add the My Home tag
                current = AddElementTag(pageUI, "MYHOME");
                AddAttribute(current, "VISIBLE", "0");

                // Add the Register tag
                current = AddElementTag(pageUI, "REGISTER");
                AddAttribute(current, "VISIBLE", "1");
                AddAttribute(current, "LINKHINT", "/Register");

                // Add the My Details tag
                current = AddElementTag(pageUI, "MYDETAILS");
                AddAttribute(current, "VISIBLE", "0");
                AddAttribute(current, "LINKHINT", "");

                // Add the Logout tag
                current = AddElementTag(pageUI, "LOGOUT");
                AddAttribute(current, "VISIBLE", "0");
                AddAttribute(current, "LINKHINT", "");
            }

            // Add the Edit page tag
            current = AddElementTag(pageUI, "EDITPAGE");
            AddAttribute(current, "VISIBLE", "0");
            AddAttribute(current, "LINKHINT", "");

            // Add the Recomended entry tag
            current = AddElementTag(pageUI, "RECOMMEND-ENTRY");
            AddAttribute(current, "VISIBLE", "0");
            AddAttribute(current, "LINKHINT", "");

            // Add the Entry Submitted tag
            current = AddElementTag(pageUI, "ENTRY-SUBBED");
            AddAttribute(current, "VISIBLE", "0");
            AddAttribute(current, "LINKHINT", "");

            // Add the Discuss tag
            current = AddElementTag(pageUI, "DISCUSS");
            AddAttribute(current, "VISIBLE", "0");
            AddAttribute(current, "LINKHINT", "");

            // Generate the random number for the banner ad?
            Random rnd = new Random((int)DateTime.Now.Ticks);
            int random = rnd.Next();

            // Add the Main Banner tag
            current = AddElementTag(pageUI, "BANNER");
            AddAttribute(current, "NAME", "main");
            AddAttribute(current, "SEED", random);
            AddAttribute(current, "SECTION", "frontpage");

            // Add the Small Banner tag
            current = AddElementTag(pageUI, "BANNER");
            AddAttribute(current, "NAME", "small");
            AddAttribute(current, "SEED", random);
            AddAttribute(current, "SECTION", "frontpage");
        }

        /// <summary>
        /// Sets the visible state and link hint for a given UI Element
        /// </summary>
        /// <param name="elementName">The name of the UI Element yoiu want to update</param>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want to update this.</param>
        /// <returns>True if the element was updated, false if not</returns>
        private bool SetElementVisibility(string elementName, bool visible, string linkHint)
        {
            // Make sure the link hint does not start with a slash
            if (linkHint != null)
            {
                linkHint = linkHint.TrimStart('/');
            }

            // Find the element to update
            XmlNode element = RootElement.SelectSingleNode("//" + elementName);
            if (element != null)
            {
                // Set the visibility flag
                if (visible)
                {
                    element.Attributes["VISIBLE"].Value = "1";
                }
                else
                {
                    element.Attributes["VISIBLE"].Value = "0";
                }

                // Update the hint
                if (linkHint != null)
                {
                    element.Attributes["LINKHINT"].Value = linkHint;
                }
            }

            // Return the verdict
            return element != null;
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetSiteHomeVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("SITEHOME", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetDontPanicVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("DONTPANIC", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetSearchVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("SEARCH", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetMyHomeVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("MYHOME", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetRegisterVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("REGISTER", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetMyDetailsVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("MYDETAILS", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetLogoutVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("LOGOUT", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetEditPageVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("EDITPAGE", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetRecommendEntryVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("RECOMMEND-ENTRY", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetEntrySubbedVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("ENTRY-SUBBED", visible, linkHint);
        }

        /// <summary>
        /// Sets the visibility for the site home element.
        /// </summary>
        /// <param name="visible">The visibility state you want to set the element to</param>
        /// <param name="linkHint">The new link hint for the element. Set this to null if you dont want it updating.</param>
        /// <returns>True if the element was updated, false if not</returns>
        public bool SetDiscussVisibility(bool visible, string linkHint)
        {
            return SetElementVisibility("DISCUSS", visible, linkHint);
        }
    }
}

File: Production/Source/Ripley/PageUI.cpp
// PageUI.cpp: implementation of the CPageUI class.
//
//////////////////////////////////////////////////////////////////////

/*

Copyright British Broadcasting Corporation 2001.

This code is owned by the BBC and must not be copied, 
reproduced, reconfigured, reverse engineered, modified, 
amended, or used in any other way, including without 
limitation, duplication of the live service, without 
express prior written permission of the BBC.

The code is provided "as is" and without any express or 
implied warranties of any kind including without limitation 
as to its fitness for a particular purpose, non-infringement, 
compatibility, security or accuracy. Furthermore the BBC does 
not warrant that the code is error-free or bug-free.

All information, data etc relating to the code is confidential 
information to the BBC and as such must not be disclosed to 
any other party whatsoever.

*/


#include "stdafx.h"
#include <time.h>
#include "PageUI.h"

#ifdef _DEBUG
#undef THIS_FILE
static char THIS_FILE[]=__FILE__;
#define new DEBUG_NEW
#endif

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

CPageUI::CPageUI(CInputContext& inputContext) : CXMLObject(inputContext)
{

}

CPageUI::~CPageUI()
{

}

/*********************************************************************************

	bool CPageUI::Initialise(CUser* pUser)

	Author:		Jim Lynn
	Created:	23/03/2000
	Inputs:		pUser - current viewing user or NULL
	Outputs:	-
	Returns:	true on success, false otherwise
	Purpose:	Initialises the UI with information from the current viewing user

*********************************************************************************/

bool CPageUI::Initialise(CUser* pUser)
{
	// First get a random number to give to the banner ad
	srand((unsigned)time(NULL));
	int iRandom = rand();
	iRandom = rand();		// Call it twice otherwise it returns the seed!

	int iUserID = 0;
	CTDVString sPassword;

	if (pUser != NULL)
	{
		pUser->GetUserID(&iUserID);
		pUser->GetPassword(sPassword);
	}
	
	// Now build the tree
	CTDVString sTreeText = "<PAGEUI>\
<SITEHOME VISIBLE='1' LINKHINT='/' />\
<SEARCH VISIBLE='1' LINKHINT='/search' />\
<DONTPANIC VISIBLE='1' LINKHINT='/dontpanic' />";
	if (iUserID > 0)
	{
		sTreeText << "<MYHOME VISIBLE='1' LINKHINT='/U" << iUserID << "'/>\n";
		sTreeText << "<REGISTER VISIBLE='0'/>\n";
		sTreeText << "<MYDETAILS VISIBLE='1' LINKHINT='/UserDetails'/>\n";
		
		// Don't display the logout button if the user hasn't defined a password
//		if (sPassword.GetLength() > 0)
//		{
			sTreeText << "<LOGOUT VISIBLE='1' LINKHINT='/Logout' />";
//		}
//		else
//		{
//			sTreeText << "<LOGOUT VISIBLE='0' />";
//		}
	}
	else
	{
		sTreeText << "<MYHOME VISIBLE='0'/>\n";
		sTreeText << "<MYDETAILS VISIBLE='0' LINKHINT=''/>\n";
		sTreeText << "<REGISTER VISIBLE='1' LINKHINT='/Register'/>\n";
		sTreeText << "<LOGOUT VISIBLE='0' LINKHINT='' />";
	}
	sTreeText << "<EDITPAGE VISIBLE='0' LINKHINT='' />";
	sTreeText << "<RECOMMEND-ENTRY VISIBLE='0' LINKHINT='' />";
	sTreeText << "<ENTRY-SUBBED VISIBLE='0' LINKHINT='' />";
	sTreeText << "<DISCUSS VISIBLE='0' LINKHINT='' />";
	sTreeText << "<BANNER NAME='main' SEED='";
	sTreeText << iRandom << "' SECTION='frontpage' />\
<BANNER NAME='small' SEED='";
	sTreeText << iRandom << "' SECTION='frontpage' />\
</PAGEUI>";

	return CreateFromXMLText(sTreeText);
}

/*********************************************************************************

	bool CPageUI::SetHomeVisibility(bool bState, const TDVCHAR *pLinkHint)

	Author:		Jim Lynn
	Created:	29/02/2000
	Inputs:		bState - true if button should be visible, false if not
				pLinkHint - URL button should go to
	Outputs:	-
	Returns:	true if succeeded, false otherwise
	Purpose:	Sets the main 'home' button (the h2g2 logo in the goo for example
				to be visible or invisible, and gives a URL hint for it.

*********************************************************************************/

bool CPageUI::SetHomeVisibility(bool bState, const TDVCHAR *pLinkHint)
{
	return SetVisibility("SITEHOME",bState, pLinkHint);
}

bool CPageUI::SetDontPanicVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("DONTPANIC",bState, pLinkHint);
}

bool CPageUI::SetSearchVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("SEARCH",bState, pLinkHint);
}

bool CPageUI::SetMyHomeVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("MYHOME",bState, pLinkHint);
}

bool CPageUI::SetRegisterVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("REGISTER",bState, pLinkHint);
}

bool CPageUI::SetMyDetailsVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("MYDETAILS",bState, pLinkHint);
}

bool CPageUI::SetLogoutVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("LOGOUT",bState, pLinkHint);
}

bool CPageUI::SetEditPageVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("EDITPAGE",bState, pLinkHint);
}

bool CPageUI::SetRecommendEntryVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("RECOMMEND-ENTRY", bState, pLinkHint);
}

bool CPageUI::SetEntrySubbedVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("ENTRY-SUBBED", bState, pLinkHint);
}


bool CPageUI::SetDiscussVisibility(bool bState, const TDVCHAR* pLinkHint)
{
	return SetVisibility("DISCUSS",bState, pLinkHint);
}

/*********************************************************************************

	bool CPageUI::SetVisibility(const TDVCHAR *pNodeName, bool bState, const TDVCHAR *pLinkHint)

	Author:		Jim Lynn
	Created:	29/02/2000
	Inputs:		pNodeName - name of node within UI structure to update
				bState - true if button should be visible, false if not
				pLinkHint - hint as to which URL should be called. Might be
				ignored by the stylesheet or changed by the transformer
	Outputs:	-
	Returns:	true if succeeded, false if failed
	Purpose:	Protected helper function to set the visibility flag and
				link hint on a given button element. Only for internal
				use.

*********************************************************************************/

bool CPageUI::SetVisibility(const TDVCHAR *pNodeName, bool bState, const TDVCHAR *pLinkHint)
{
	CTDVString hint = pLinkHint;
	if (hint.GetLength() > 1 && hint.GetAt(0) == '/')
	{
		hint.RemoveLeftChars(1);
	}

	bool bSuccess = true;
	
	// Find the node we're after
	CXMLTree* pNode = m_pTree->FindFirstTagName(pNodeName);
	if (pNode == NULL)
	{
		bSuccess = false;
	}
	else
	{
		// Set the VISIBLE attribute to reflect the state flag
		if (bState)
		{
			pNode->SetAttribute("VISIBLE","1");
		}
		else
		{
			pNode->SetAttribute("VISIBLE","0");
		}
		// Set the LINKHINT attribute if passed to us
		if (pLinkHint != NULL)
		{
			pNode->SetAttribute("LINKHINT", hint);
		}
		bSuccess = true;
	}
	return bSuccess;
}
